lib/diyepw/diyepw/get_noaa_isd_lite_file.py

```diff
@@ -0,0 +1,90 @@
+import urllib.request as _request
+import re as _re
+import os as _os
+import pandas as _pd
+
+def get_noaa_isd_lite_file(wmo_index:int, year:int, output_dir:str = None, force_update = False) -> str:
+    """
+    Given a WMO index and a year, retrieve the corresponding NOAA ISD Lite AMY file
+    :param wmo_index:
+    :param year:
+    :param output_dir: Optional output directory - if not specified, the file will be saved to a package directory.
+        If the directory already contains a NOAA ISD Lite file matching the requested WMO Index and year, then a new
+        file will not be downloaded from NOAA and that file's path will be returned
+    :param force_update: Pass True to force a new ISD Lite file to be downloaded, even if it already exists in the
+       output directory.
+    :return: The path to the NOAA ISD Lite file
+    """
+    if output_dir is None:
+        this_dir = _os.path.dirname(_os.path.realpath(__file__))
+        output_dir = _os.path.join(this_dir, 'files', 'noaa_isd_lite_files')
+
+    if not _os.path.exists(output_dir):
+        _os.mkdir(output_dir)
+
+    # On the NOAA website, the ISD Lite files are named with a third number between WMO and year, but
+    # since we don't use that third number for anything and it complicates identifying a file for a
+    # WMO/Year combination, we simplify the name to only contain the values we care about
+    file_name = f"{wmo_index}-{year}.gz"
+    file_path = _os.path.join(output_dir, file_name)
+
+    # Download the ISD Lite file if it's not already in the output directory
+    if force_update or not _os.path.exists(file_path):
+        url = _get_noaa_isd_lite_file_url(year, wmo_index)
+        with _request.urlopen(url) as response:
+            with open(file_path, 'wb') as downloaded_file:
+                downloaded_file.write(response.read())
+
+    return file_path
+
+def _get_noaa_isd_lite_file_url(year:int, wmo_index:int) -> str:
+    catalog = _get_noaa_isd_lite_file_catalog(year)
+    file_name = list(catalog.loc[catalog['wmo_index'] == wmo_index]['file_name'])[0]
+    return f"https://www1.ncdc.noaa.gov/pub/data/noaa/isd-lite/{year}/{file_name}"
```

Comment:
websites break.  This should be raised up as a config option default

Reply:
It's not at all clear to me how this could work. @amandadsmith and I discussed this a fair bit before I started developing this function, and there's really no sensible way I can think of to support a configurable option for what URL to retrieve NOAA's ISD Lite files from. This is just where those files are stored, and we're providing the ability to automatically download them because we can't legally include them as part of the package's built-in data because NOAA asserts copyright over the files. It's definitely true that websites break, which was a concern I raised in that discussion, but our options as I see them are to use this URL and be prepared to release updates in response to any changes, or license the files and include them in the package.

Reply:
I am not at my computer (on phone) so I cannot check the details. But I don't like stuff being downloaded in packages I use because the user has no control over something that is being brought onto their machine. I recommend setting the auto download as an option by the user with the default to False, with an additional option to provide a directory containing the files they may have already downloaded. Expose the fetch function to the user so that the may download and archive where they have write access to do so and where they may want to keep the data. This data will need to be sited when using products built from it. To do so, we would need the appropriate metadata such as date accessed, the license you mention, etc. 

Also raise informative exceptions when the website breaks.

Reply:
I have created https://github.com/IMMM-SFA/diyepw/issues/28 to track your request to change how downloads of TMY3 and ISD Lite files work. I've proposed a compromise there that I think is a bit more user friendly than a configuration option, but still allows the user full control over whether files are downloaded. I'd like to make sure that you and Amanda are both on board with whatever solution we go with before I proceed.

I have created https://github.com/IMMM-SFA/diyepw/issues/30 to make sure we produce sufficiently informative error messages in the event that the EnergyPlus or NOAA sites go down.

Both issues will be developed in the csv_based_script branch.